src/mame/nabu/nabupc.cpp

```diff
@@ -0,0 +1,509 @@
+
+	rs232_port_device &kbd(RS232_PORT(config, "kbd", keyboard_devices, "nabu"));
+	kbd.rxd_handler().set(m_kbduart, FUNC(i8251_device::write_rxd));
+
+	// HCCA
+	AY31015(config, m_hccauart);
+	m_hccauart->set_auto_rdav(true);
+	m_hccauart->write_dav_callback().set(FUNC(nabupc_state::hcca_dr_w));
+	m_hccauart->write_tbmt_callback().set(FUNC(nabupc_state::hcca_tbre_w));
+	m_hccauart->write_fe_callback().set(FUNC(nabupc_state::hcca_fe_w));
+	m_hccauart->write_or_callback().set(FUNC(nabupc_state::hcca_oe_w));
+	m_hccauart->write_so_callback().set("hcca", FUNC(rs232_port_device::write_txd));
+
+	rs232_port_device &hcca(RS232_PORT(config, "hcca", hcca_devices, "pty"));
+	hcca.rxd_handler().set(m_hccauart, FUNC(ay31015_device::write_si));
+
+	// Printer
+	output_latch_device &prndata(OUTPUT_LATCH(config, "prndata"));
+	CENTRONICS(config, m_centronics, printer_devices, nullptr);
+	m_centronics->set_output_latch(prndata);
+	m_centronics->busy_handler().set(FUNC(nabupc_state::centronics_busy_handler));
+
+	// Clocks
+	clock_device &sclk(CLOCK(config, "sclk", 10.738635_MHz_XTAL / 96)); // 111.9 kHz Clock
+	sclk.signal_handler().set(m_kbduart, FUNC(i8251_device::write_rxc));
+
+	clock_device &pclk(CLOCK(config, "pclk", 10.738635_MHz_XTAL / 6));  // 1.79 Mhz Clock
+	pclk.signal_handler().set(m_hccauart, FUNC(ay31015_device::write_rcp));
+	pclk.signal_handler().append(m_hccauart, FUNC(ay31015_device::write_tcp));
+
+	// Bus
+	NABUPC_OPTION_BUS(config, m_bus, 10.738635_MHz_XTAL / 3);
+	m_bus->out_int_callback<0>().set(FUNC(nabupc_state::j9_int_w));
+	m_bus->out_int_callback<1>().set(FUNC(nabupc_state::j10_int_w));
+	m_bus->out_int_callback<2>().set(FUNC(nabupc_state::j11_int_w));
+	m_bus->out_int_callback<3>().set(FUNC(nabupc_state::j12_int_w));
+	NABUPC_OPTION_BUS_SLOT(config, "option1", m_bus, 0, bus::nabupc::option_bus_devices, nullptr);
+	NABUPC_OPTION_BUS_SLOT(config, "option2", m_bus, 1, bus::nabupc::option_bus_devices, nullptr);
+	NABUPC_OPTION_BUS_SLOT(config, "option3", m_bus, 2, bus::nabupc::option_bus_devices, nullptr);
+	NABUPC_OPTION_BUS_SLOT(config, "option4", m_bus, 3, bus::nabupc::option_bus_devices, nullptr);
+}
+
+
+//**************************************************************************
+//  MACHINE INITIALIZATION
+//**************************************************************************
+
+// Machine Start
+void nabupc_state::machine_start()
+{
+	m_leds.resolve();
+
+	m_hccauart->write_np(1);
+	m_hccauart->write_nb2(1);
+	m_hccauart->write_nb1(1);
+	m_hccauart->write_eps(0);
+	m_hccauart->write_tsb(0);
+	m_hccauart->write_cs(1);
+	m_hccauart->write_swe(0);
+
+	save_item(NAME(m_irq_in_prio));
+	save_item(NAME(m_int_lines));
+	save_item(NAME(m_porta));
+	save_item(NAME(m_portb));
+	save_item(NAME(m_control));
+	save_item(NAME(m_bios_size));
+}
+
+// Machine Reset
+void nabupc_state::machine_reset()
+{
+	m_irq_in_prio = 0xFF;
+	m_int_lines = 0;
+	m_porta = 0;
+	m_portb = 0;
```

Comment:
`m_portb` represents input lines from the Centronics port – you shouldn’t reset the values of inputs on device reset, you should let the device on the other side of the connector update the inputs if it responds to reset.